src/topp/MultiplexResolver.cpp

```diff
@@ -0,0 +1,674 @@
+   * @param theoretical_patterns    list of theoretical delta mass patterns
+   * @param delta_mass_matched    Was this delta mass in the theoretical pattern matched?
+   * @param index_label_set    index within the pattern at which the label sets were matched
+   * 
+   * @return index of matching pattern
+   */
+  int findMatchingPattern(const ConsensusMap::ConstIterator consensus, const MultiplexDeltaMasses::LabelSet& label_set, const std::vector<MultiplexDeltaMasses>& theoretical_patterns, std::vector<bool>& delta_mass_matched, int& index_label_set)
+  {
+    // loop over theoretical patterns
+    for (std::vector<MultiplexDeltaMasses>::const_iterator it_pattern = theoretical_patterns.begin(); it_pattern != theoretical_patterns.end(); ++it_pattern)
+    {
+      std::vector<MultiplexDeltaMasses::DeltaMass> pattern = it_pattern->getDeltaMasses();
+      
+      double shift = matchLabelSet(pattern, label_set, index_label_set);
+      if (!boost::math::isnan(shift))
+      {        
+        // reset boolean vector
+        unsigned i = delta_mass_matched.size();
+        delta_mass_matched.assign(i,false);
+        
+        bool match = matchDeltaMasses(consensus, pattern, shift, delta_mass_matched);
+        if (match)
+        {
+          return (it_pattern - theoretical_patterns.begin());
+        }
+      }
+    }
+    
+    return -1;
+  }
+  
+  /**
+   * @brief find the m/z for the complete consensus
+   *
+   * @param mz    m/z of the incomplete consensus
+   * @param charge    charge of the incomplete consensus
+   * @param pattern    matching theoretical delta mass pattern
+   * @param delta_mass_matched    Was this delta mass in the theoretical pattern matched?
+   * 
+   * @return m/z for the complete consensus
+   */
+  double findNewMZ(double mz, int charge, const std::vector<MultiplexDeltaMasses::DeltaMass>& pattern, const std::vector<bool>& delta_mass_matched)
+  {
+    // loop over delta masses in theoretical pattern
+    std::vector<MultiplexDeltaMasses::DeltaMass>::const_iterator it_mass_shift;
+    std::vector<bool>::const_iterator it_delta_mass_matched;
+    for (it_mass_shift = pattern.begin(), it_delta_mass_matched = delta_mass_matched.begin();
+         it_mass_shift != pattern.end(), it_delta_mass_matched != delta_mass_matched.end();
+         ++it_mass_shift, ++it_delta_mass_matched)
+    {
+      // find the first match
+      if (*it_delta_mass_matched)
+      {
+        return (mz * charge - it_mass_shift->delta_mass)/charge;
+      }
+    }
+    
+    // Should never happen.
+    return mz;
+  }
+  
+  /**
+   * @brief complete consensus
+   *
+   * @param consensus    (possibly) incomplete consensus
+   * @param pattern    matching theoretical delta mass pattern
+   * @param delta_mass_matched    Was this delta mass in the theoretical pattern matched?
+   * @param index_label_set    index within the pattern at which the label sets were matched
```

Comment:
here the parameter is not used as return parameter. This is confusing as the same one has been used as return parameter instead. Another reason why to return a pair (as mentioned above).